[DPP-954][Docs] Improve pruning offsets docs (#16254)
digital-asset/daml: docs/source/ops/pruning.rst

- Pruning is potentially a long-running operation and demanding one in terms of system resources; as such, it may significantly reduce Daml Ledger API throughput and increase latency while it is being performed. It is thus strongly recommended to plan pruning invocations, preferably, when the system is offline or at least when very low system utilization is expected.
- Pruning may degrade the behavior of or abort in-progress requests if the pruning offset is too recent. In particular, the system might misbehave if command completions are pruned before the command trackers are able to process the completions.

> - Command deduplication and command tracker retention should always be configured in such a way, that the associated windows don't overlap with the pruning window, so that their operation is unaffected by pruning.

Reviewer: ```suggestion
- Command deduplication and command tracker retention should always be configured so that the associated windows don't overlap with the pruning window to ensure that their operation is unaffected by pruning.
```

- Pruning may affect the behavior of Ledger API calls that allow to read data from the ledger: see the next sub-section for more information about API impacts.
- Pruning of all divulged contracts (see :ref:`Prune Request <com.daml.ledger.api.v1.admin.PruneRequest>`) does not preserve application visibility over contracts divulged up to the pruning offset, hence applications making use of pruned divulged contracts might start experiencing failed command submissions: see the section below for determining a suitable pruning offset.

.. warning::
  Participants may know of contracts for which they don't know the current activeness status. This happens through :ref:`divulgence <da-model-divulgence>` where a party learns of the existence of a contract without being guaranteed to ever see its archival. Such contracts are pruned by the feature described on this page as not doing so could easily lead to an ever growing participant state.

> 

Reviewer 2: hopefully this new line doesn't extract the subsequent sentences from the warning section

Author: Check and it looks fine. The warning and at " lead to an ever growing participant state."

Determine a Suitable Pruning Offset
-----------------------------------

> The :ref:`Transaction Service <transaction-service>` and the :ref:`Active Contract Service <active-contract-service>` provide offsets of the ledger end of the Transactions, and of Active Contracts snapshots respectively. Such offsets can be passed unchanged to `prune` calls, as long as they are lexicographically lower than the current ledger end. An additional constraint imposed by Canton (subject to future improvements) is that the participant you are pruning must have already exchanged the ACS commitments with other participants for the offset that you prune at.

Reviewer 2: Not sure how helpful it is to mention that. But it could be setting some unnecessary expectations.
```suggestion
The :ref:`Transaction Service <transaction-service>` and the :ref:`Active Contract Service <active-contract-service>` provide offsets of the ledger end of the Transactions, and of Active Contracts snapshots respectively. Such offsets can be passed unchanged to `prune` calls, as long as they are lexicographically lower than the current ledger end. An additional constraint imposed by Canton is that the participant you are pruning must have already exchanged the ACS commitments with other participants for the offset that you prune at.
```

Reviewer 2: I think we are over-detailing the acs commitment stuff without saying where should the user look to understand better. I would actually reduce the explanation and just reference a Canton docs page (@oliverse-da do we have pruning ops page mentioning the safe to prune offset for Canton)?

Author: We have this: 
> [Current Limitations](https://www.canton.io/docs/dev/user-manual/usermanual/pruning.html#current-limitations)
> ..
> Pruning of participants requires the participant to have received a commitment from each counter-participant with which it shares a contract. If a participant becomes defunct and stops sending commitments, pruning of the participant will not work. Therefore, before you disable a participant, please make sure that is not involved in any contract. This is a current limitation.

Reviewer 3: Thanks for digging up this page @pbatko-da. That is indeed the only thing we state about participant commitments on the canton pruning page.

We used to have a whole separate section on pruning in test-environments where users used to be confronted with issues pruning "more current state", but I removed that as part of switching customers to automatic background pruning.

I am fine either way with leaving the detail here and / or add a link to this pruning page.

Author: FYI Added a link to https://www.canton.io/docs/dev/user-manual/usermanual/pruning.html#current-limitations

Reviewer 2: alright, since this is not documented elsewhere, fine to leave this as is wrt to the description of the ACS commitments then. I would still add a link to that page that you found 👍

Reviewer 3: Not sure I am up on the latest regarding documentation links - shouldn't the link point to https://docs.daml.com/ops/pruning.html rather than the canton-io page?

Reviewer: Correct. And relative too.

Author: What I'm editing in this PR is the same page so I think there's no point to link to itself. Will remove the link

Author: I made the link to be `<../canton/usermanual/pruning.html>` which I hope will point to https://docs.daml.com/canton/usermanual/pruning.html when docs are released. (Is there an easy way to test it locally? I guess not)
